python/lsst/pipe/base/graph/_implDetails.py

```diff
@@ -181,3 +208,77 @@ def __contains__(self, key: _T) -> bool:
             Boolean of the presence of the supplied key
         """
         return key in self._producers or key in self._consumers
+
+
+def _pruner(datasetRefDict: _DatasetTracker[DatasetRef, QuantumNode], refs: Iterable[DatasetRef], *,
+            prunes: Optional[Set[QuantumNode]] = None):
+    if prunes is None:
+        prunes = set()
+    for ref in refs:
+        # make a copy here, because this structure will be modified in
+        # recursion, hitting a node more than once wont be much of an
+        # issue, as we skip anything that has been processed
+        nodes = set(datasetRefDict.getConsumers(ref))
+        for node in nodes:
+            # This node will never be associated with this ref
+            datasetRefDict.removeConsumer(ref, node)
+            if node in prunes:
+                continue
+            # find the connection corresponding to the input ref
+            connectionRefs = node.quantum.inputs.get(ref.datasetType)
+            # Ref must be an initInput ref and we want to ignore those
+            if connectionRefs is None:
+                continue
+            tmpRefs = set(connectionRefs).difference((ref,))
+            tmpConnections = NamedKeyDict(node.quantum.inputs.items())
+            tmpConnections[ref.datasetType] = list(tmpRefs)
+            helper = AdjustQuantumHelper(inputs=tmpConnections, outputs=node.quantum.outputs)
+            # assert to make the type checker happy, it should not actually be
+            # possible to not have dataId set to None at this point
+            assert node.quantum.dataId is not None
+            # Try to adjust the quantum with the reduced refs to make sure the
+            # node will still satisfy all its conditions.
+            #
+            # If it cant because NoWorkFound is raise, that means a connection
+            # is no longer present, and the node should be removed from the
+            # graph
+            try:
+                helper.adjust_in_place(node.taskDef.connections, node.taskDef.label, node.quantum.dataId)
+                newQuantum = Quantum(taskName=node.quantum.taskName, taskClass=node.quantum.taskClass,
+                                     dataId=node.quantum.dataId, initInputs=node.quantum.initInputs,
+                                     inputs=helper.inputs, outputs=helper.outputs)
+                # If the inputs or outputs were adjusted than the graph builder
+                # thought, dissassociate node from those refs, and if they are
+                # output refs, prune them from downstream tasks
+                for condition, existingMapping, newMapping, remover in (
+                        (helper.inputs_adjusted, node.quantum.inputs, helper.inputs,
+                         datasetRefDict.removeConsumer),
+                        (helper.outputs_adjusted, node.quantum.outputs, helper.outputs,
+                         datasetRefDict.removeProducer)):
```

Comment:
I bet this would overall be a lot easier to read if you instead modified `AdjustQuantumHelper` to remember the datasets that were removed directly (it already has a similar loop to compute its `bool` flags; it just throws what you want away).

Fine to defer this to another ticket if you've already done a lot of one-off testing that you don't think you have time to repeat.

Reply:
Indeed I have, I think this can stand until another ticket or some of the future reworks. This really should not be triggered right now, and even when it does it will be really rare, so no rush.